File: app/src/main/java/daocacao/myprecious/ui/MonthActivity.java
package daocacao.myprecious.ui;

import android.content.DialogInterface;
import android.os.Bundle;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import daocacao.myprecious.AppLoader;
import daocacao.myprecious.adapters.CalendarRecyclerAdapter;
import daocacao.myprecious.R;
import daocacao.myprecious.data.Day;
import daocacao.myprecious.data.Gold;
import io.realm.RealmResults;

public class MonthActivity extends AppCompatActivity {

    TextView txtMonthTitle;
    ImageView btnLeft, btnRight;
    RecyclerView recyclerView;
    TextView txtTotal;
    FloatingActionButton fab;

    CalendarRecyclerAdapter adapter;

    DayFragment dayFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initViews();

        dayFragment = new DayFragment();

        adapter = new CalendarRecyclerAdapter(this);

        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new GridLayoutManager(this, 7));

        btnLeft.setOnClickListener(v -> updateCalendar(-1));
        btnRight.setOnClickListener(v -> updateCalendar(1));

        fab.setOnClickListener(v -> dayFragment.addNewGoldDialog());
        fab.hide();

        updateCalendar(0);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        showEditLimitsDialog();
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        showTotalInMonth();
        adapter.notifyDataSetChanged();
        fab.hide();
    }

    private void initViews() {
        setContentView(R.layout.month_activity);

        txtMonthTitle = (TextView) findViewById(R.id.txt_month_title);
        btnLeft = (ImageView) findViewById(R.id.btn_left);
        btnRight = (ImageView) findViewById(R.id.btn_right);
        recyclerView = (RecyclerView) findViewById(R.id.recycler_view);
        txtTotal = (TextView) findViewById(R.id.txt_total);
        fab = (FloatingActionButton) findViewById(R.id.fab);
    }

    private void updateCalendar(int count) {
        if (count != 0) AppLoader.updateCalendar(count);

        List<Date> cells = new ArrayList<>();
        Calendar calendar = (Calendar) AppLoader.currentCalendar.clone();
        calendar.set(Calendar.DAY_OF_MONTH, 1);

        int monthBeginningCell = calendar.get(Calendar.DAY_OF_WEEK) - 1;
        txtMonthTitle.setText(new SimpleDateFormat("MMM yyyy", Locale.getDefault()).format(calendar.getTime()));

        calendar.add(Calendar.DAY_OF_MONTH, -monthBeginningCell);

        while (cells.size() < AppLoader.DAYS_COUNT) {
            cells.add(calendar.getTime());
            calendar.add(Calendar.DAY_OF_MONTH, 1);
        }
        adapter.updateData(cells);
        showTotalInMonth();
    }

    private void showTotalInMonth() {
        float total = 0;
        int currentMonth = AppLoader.currentDate[1];
        int currentYear = AppLoader.currentDate[2];
        RealmResults<Day> days = AppLoader.realm.where(Day.class)
                .equalTo("year", currentYear)
                .equalTo("month", currentMonth)
                .findAll();

        for (Day day : days)
            if (!day.getGolds().isEmpty())
                for (Gold gold : day.getGolds())
                    total += gold.getPrice();

        txtTotal.setText(String.format(Locale.getDefault(), "%s %.02f", "Spended:", total));
    }

    public void showDay(String date) {
        dayFragment.setDate(date);
        getSupportFragmentManager()
                .beginTransaction()
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                .replace(R.id.container_layout, dayFragment)
                .addToBackStack("gold_frag")
                .commit();
        fab.show();
    }

    private void showEditLimitsDialog() {
        View view = LayoutInflater.from(this).inflate(R.layout.add_gold_layout, new LinearLayout(this));
        EditText edDayTotal = (EditText) view.findViewById(R.id.ed_first);
        EditText edMonthTotal = (EditText) view.findViewById(R.id.ed_second);

        edDayTotal.setHint(R.string.txt_day_limit);
        edMonthTotal.setVisibility(View.GONE);

        AlertDialog alertDialog = new AlertDialog.Builder(this)
                .setView(view)
                .setPositiveButton("Ok", null)
                .create();

        alertDialog.setOnShowListener(dialog ->
                alertDialog.getButton(DialogInterface.BUTTON_POSITIVE).setOnClickListener(v -> {

                    if (!TextUtils.isEmpty(edDayTotal.getText())) {
                        float dayLimit = Float.parseFloat(edDayTotal.getText().toString());
                        AppLoader.savePrefs(dayLimit);
                        dialog.dismiss();
                        adapter.notifyDataSetChanged();
                        showTotalInMonth();
                        if (dayFragment.isVisible()) dayFragment.setTotal();
                    } else
                        edDayTotal.startAnimation(AnimationUtils.loadAnimation(this, R.anim.shake_txt_anim));
                }));

        alertDialog.show();
    }
}

File: app/src/main/java/daocacao/myprecious/ui/MonthActivity.kt
package daocacao.myprecious.ui

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import daocacao.myprecious.AppPrefs
import daocacao.myprecious.CalendarManager
import daocacao.myprecious.DialogManager
import daocacao.myprecious.R
import daocacao.myprecious.adapters.days.DaysAdapter
import daocacao.myprecious.realm.RealmManager
import kotlinx.android.synthetic.main.activity_month.btn_left
import kotlinx.android.synthetic.main.activity_month.btn_right
import kotlinx.android.synthetic.main.activity_month.recycler_view
import kotlinx.android.synthetic.main.activity_month.txt_month_title
import kotlinx.android.synthetic.main.activity_month.txt_total
import java.util.*

class MonthActivity : AppCompatActivity() {

    private val adapter by lazy { DaysAdapter(this) }

    private val prefs by lazy { AppPrefs(this) }
    private val realmManager by lazy { RealmManager(this) }
    private val calendarManager by lazy { CalendarManager() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_month)

        recycler_view.adapter = adapter

        adapter.calendarManager = calendarManager
        adapter.dayLimit = prefs.dayLimit
        adapter.onDayClick = this::showDay

        btn_left.setOnClickListener { calendarManager.getPrevMonth(this::updateView) }
        btn_right.setOnClickListener { calendarManager.getNextMonth(this::updateView) }

        calendarManager.getCurrentMonth(this::updateView)

        realmManager.observable
                .subscribe {
                    adapter.update(it)
                    showTotalInMonth()
                }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        DialogManager(this).showEditLimitsDialog {
            prefs.dayLimit = it
            adapter.dayLimit = it
            adapter.notifyDataSetChanged()
            showTotalInMonth()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun updateView(title: String, days: ArrayList<Date>) {
        adapter.cells = days
        txt_month_title.text = title
        showTotalInMonth()
    }

    private fun showTotalInMonth() {
        val currentMonth = calendarManager.selectedDate[1]
        val currentYear = calendarManager.selectedDate[2]

        val month = realmManager.getMonth(currentMonth, currentYear)

        val total = month.map { it.records.map { it.spended }.sum() }.sum()
        val average = month.map { it.records.map { it.spended }.average() }.average()

        txt_total.text = String.format(Locale.getDefault(), "Spended: %.02f (%.02f)", total, average)
    }

    private fun showDay(date: String) {
        DayFragment.getInstance(date).apply {
            show(supportFragmentManager, this.javaClass.canonicalName)
        }
    }
}